core/src/main/java/com/microsoft/applicationinsights/internal/config/connection/ConnectionString.java

```diff
@@ -0,0 +1,192 @@
+    static class KeyValuePairScanner implements Iterator<Map.Entry<String, String>> {
+
+        enum State {
+            KEY,
+            VALUE,
+            FINISHED,
+        }
+
+        private final String input;
+        private final char pairSeparator;
+        private final char valueSeparator;
+
+        private Map.Entry<String, String> cachedNext = null;
+        private int startIndex = 0;
+        private State state = State.KEY;
+
+        KeyValuePairScanner(String input, char pairSeparator, char valueSeparator) {
+            this.input = input;
+            this.pairSeparator = pairSeparator;
+            this.valueSeparator = valueSeparator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return cachedNext != null || cacheNext();
+        }
+
+        @Override
+        public Map.Entry<String, String> next() {
+            if (!hasNext()) {
+                return null;
+            }
+            Map.Entry<String, String> result = cachedNext;
+            cachedNext = null;
+            return result;
+        }
+
+        /**
+         *
+         * @return true if an entry was cached.
+         */
+        private boolean cacheNext() {
+            if (state == State.FINISHED) {
+                return false;
+            }
+
+            // TODO handle error cases
+
+            char separator = valueSeparator;
+            int i = startIndex;
+            for (; i < input.length(); i++) {
+                if (input.charAt(i) == separator) {
+                    if (state == State.KEY) {
+                        cachedNext = new SimpleEntry<>(input.substring(startIndex, i), null);
+                        startIndex = ++i; // skip separator
+                        separator = pairSeparator;
+                        state = State.VALUE;
+                    } else if (state == State.VALUE) {
+                        cachedNext.setValue(input.substring(startIndex, i));
+                        startIndex = ++i; // skip separator
+                        state = State.KEY;
+                        break;
+                    }
+                }
+            }
+
+            if (i == input.length()) {
+                cachedNext.setValue(input.substring(startIndex, i));
+                state = State.FINISHED;
+            }
+
+            return true;
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public static ConnectionConfiguration parse(String connectionString) throws ConnectionStringParseException {
+        // parse key value pairs
+        final Map<String, String> kvps;
+        try {
+            kvps = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+            kvps.putAll(Splitter.on(';').trimResults().omitEmptyStrings().withKeyValueSeparator('=').split(connectionString));
+        } catch (IllegalArgumentException e) {
+            throw new InvalidConnectionStringException(e);
```

Comment:
I think add message, e.g. "Error parsing connection string: ...", and include problematic connection string (in case it's picking up connection string that user didn't expect.

Reply:
Sounds good:

```
com.microsoft.applicationinsights.internal.config.connection.InvalidConnectionStringException: Error parsing connection string: "Authorization=ikey;InstrumentationKey==fake-ikey"
	at com.microsoft.applicationinsights.internal.config.connection.ConnectionString.parseInto(ConnectionString.java:21)
	at com.microsoft.applicationinsights.internal.config.connection.ConnectionStringTests.multipleKeySeparatorsIsInvalid(ConnectionStringTests.java:194)
	...
Caused by: java.lang.IllegalArgumentException: Chunk [InstrumentationKey==fake-ikey] is not a valid entry
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:217)
	at com.google.common.base.Splitter$MapSplitter.split(Splitter.java:513)
	at com.microsoft.applicationinsights.internal.config.connection.ConnectionString.parseInto(ConnectionString.java:19)
```